File: ls/server/src/test/java/org/opencds/cqf/cql/ls/server/LanguageServerTest.java
package org.opencds.cqf.cql.ls.server;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertNotNull;
import static org.junit.jupiter.api.Assertions.assertNull;

import org.eclipse.lsp4j.Hover;
import org.eclipse.lsp4j.HoverParams;
import org.eclipse.lsp4j.MarkupContent;
import org.eclipse.lsp4j.Position;
import org.eclipse.lsp4j.TextDocumentIdentifier;
import org.eclipse.lsp4j.services.LanguageClient;
import org.junit.jupiter.api.Test;
import org.mockito.Mockito;
import org.opencds.cqf.cql.ls.server.config.TestConfig;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.context.SpringBootTest;

// Just sketching out some tests here to get an idea of the scaffolding we need to make the language
// server easily testable
// We'll need to split out a few components to make it easier.

@SpringBootTest(classes = {TestConfig.class})
class LanguageServerTest {

    @Autowired
    CqlLanguageServer server;

    LanguageClient client = Mockito.mock(LanguageClient.class);

    @Test
    void handshake() throws Exception {
        // https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#lifeCycleMessages

        // Sequence for initialization
        // initialize
        // initialized
        // ...
        // do language server stuff
        // ...
        // shutdown
        // exit

        assertNotNull(server);
    }

    @Test
    void hoverInt() throws Exception {
        Hover hover = server.getTextDocumentService()
                .hover(new HoverParams(
                        new TextDocumentIdentifier("/org/opencds/cqf/cql/ls/server/Two.cql"), new Position(5, 2)))
                .get();

        assertNotNull(hover);
        assertNotNull(hover.getContents().getRight());

        MarkupContent markup = hover.getContents().getRight();
        assertEquals("markdown", markup.getKind());
        assertEquals("```cql\nSystem.Integer\n```", markup.getValue());
    }

    @Test
    void hoverNothing() throws Exception {
        Hover hover = server.getTextDocumentService()
                .hover(new HoverParams(
                        new TextDocumentIdentifier("/org/opencds/cqf/cql/ls/server/Two.cql"), new Position(2, 0)))
                .get();

        assertNull(hover);
    }

    @Test
    void hoverList() throws Exception {
        Hover hover = server.getTextDocumentService()
                .hover(new HoverParams(
                        new TextDocumentIdentifier("/org/opencds/cqf/cql/ls/server/Two.cql"), new Position(8, 2)))
                .get();

        assertNotNull(hover);
        assertNotNull(hover.getContents().getRight());

        MarkupContent markup = hover.getContents().getRight();
        assertEquals("markdown", markup.getKind());
        assertEquals("```cql\nlist<System.Integer>\n```", markup.getValue());
    }
}

File: ls/server/src/test/java/org/opencds/cqf/cql/ls/server/LanguageServerTest.kt
package org.opencds.cqf.cql.ls.server

import org.eclipse.lsp4j.HoverParams
import org.eclipse.lsp4j.Position
import org.eclipse.lsp4j.TextDocumentIdentifier
import org.eclipse.lsp4j.services.LanguageClient
import org.greenrobot.eventbus.EventBus
import org.greenrobot.eventbus.Logger.JavaLogger
import org.junit.jupiter.api.Assertions.assertEquals
import org.junit.jupiter.api.Assertions.assertNotNull
import org.junit.jupiter.api.Assertions.assertNull
import org.junit.jupiter.api.BeforeAll
import org.junit.jupiter.api.Test
import org.opencds.cqf.cql.ls.server.manager.CompilerOptionsManager
import org.opencds.cqf.cql.ls.server.manager.CqlCompilationManager
import org.opencds.cqf.cql.ls.server.manager.IgContextManager
import org.opencds.cqf.cql.ls.server.plugin.CommandContribution
import org.opencds.cqf.cql.ls.server.provider.FormattingProvider
import org.opencds.cqf.cql.ls.server.provider.HoverProvider
import org.opencds.cqf.cql.ls.server.service.CqlTextDocumentService
import org.opencds.cqf.cql.ls.server.service.CqlWorkspaceService
import org.opencds.cqf.cql.ls.server.service.TestContentService
import java.util.concurrent.CompletableFuture

class LanguageServerTest {

    companion object {
        private lateinit var server: CqlLanguageServer

        @BeforeAll
        @JvmStatic
        fun beforeAll() {
            val eventBus = EventBus.builder().logger(JavaLogger("eventBus")).build()
            val cs = TestContentService()
            val compilationManager = CqlCompilationManager(cs, CompilerOptionsManager(cs), IgContextManager(cs))
            val languageClientFuture = CompletableFuture<LanguageClient>()
            val commandsFuture = CompletableFuture<List<CommandContribution>>()
            commandsFuture.complete(emptyList())
            server = CqlLanguageServer(
                languageClientFuture,
                CqlWorkspaceService(languageClientFuture, commandsFuture, mutableListOf(), eventBus),
                CqlTextDocumentService(languageClientFuture, HoverProvider(compilationManager), FormattingProvider(cs), eventBus)
            )
        }
    }

    @Test
    fun handshake() {
        assertNotNull(server)
    }

    @Test
    fun hoverInt() {
        val hover = server.getTextDocumentService()
            .hover(HoverParams(TextDocumentIdentifier("/org/opencds/cqf/cql/ls/server/Two.cql"), Position(5, 2)))
            .get()

        assertNotNull(hover)
        assertNotNull(hover!!.contents.right)

        val markup = hover.contents.right
        assertEquals("markdown", markup.kind)
        assertEquals("```cql\nSystem.Integer\n```", markup.value)
    }

    @Test
    fun hoverNothing() {
        val hover = server.getTextDocumentService()
            .hover(HoverParams(TextDocumentIdentifier("/org/opencds/cqf/cql/ls/server/Two.cql"), Position(2, 0)))
            .get()

        assertNull(hover)
    }

    @Test
    fun hoverList() {
        val hover = server.getTextDocumentService()
            .hover(HoverParams(TextDocumentIdentifier("/org/opencds/cqf/cql/ls/server/Two.cql"), Position(8, 2)))
            .get()

        assertNotNull(hover)
        assertNotNull(hover!!.contents.right)

        val markup = hover.contents.right
        assertEquals("markdown", markup.kind)
        assertEquals("```cql\nlist<System.Integer>\n```", markup.value)
    }
}
